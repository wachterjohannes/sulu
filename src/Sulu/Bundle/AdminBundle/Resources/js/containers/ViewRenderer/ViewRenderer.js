// @flow
import React from 'react';
import type {Element} from 'react';
import {observer} from 'mobx-react';
import Router from '../../services/Router';
import type {Route} from '../../services/Router';
import viewRegistry from './registries/ViewRegistry';
import {sidebarStore} from '../Sidebar';

type Props = {
    router: Router,
};

@observer
export default class ViewRenderer extends React.Component<Props> {
    getKey = (route: Route) => {
        if (!route.rerenderAttributes) {
            return undefined;
        }

        const {
            router: {
                attributes,
            },
        } = this.props;

        const rerenderAttributeValues = [];

        route.rerenderAttributes.forEach((rerenderAttribute) => {
            if (attributes.hasOwnProperty(rerenderAttribute)) {
                rerenderAttributeValues.push(attributes[rerenderAttribute]);
            }
        });

        return rerenderAttributeValues.join('__');
    };

    renderView(route: Route, child: Element<*> | null = null, , hasSidebar: boolean = false) {
        const {router} = this.props;
        const {view} = route;

        const View = viewRegistry.get(view);
        if (!View) {
            throw new Error('View "' + view + '" has not been found');
        }

        // if any view has sidebar - the sidebar will be visible
        hasSidebar = View.hasSidebar ? true : hasSidebar;

        const element = (
            <View router={router} route={route} key={this.getKey(route)}>
                {(props) => child ? React.cloneElement(child, props) : null}
            </View>
        );

        if (!route.parent) {
            if (!hasSidebar) {
                sidebarStore.clearConfig();
            }

            return element;
        }

        return this.renderView(route.parent, element, hasSidebar);
    }

    render() {
        return this.renderView(this.props.router.route);
    }
}
