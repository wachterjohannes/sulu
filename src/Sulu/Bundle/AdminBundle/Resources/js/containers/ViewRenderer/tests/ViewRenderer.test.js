/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import {render, shallow} from 'enzyme';
import ViewRenderer from '../ViewRenderer';
import viewRegistry from '../registries/ViewRegistry';
import sidebarStore from '../../Sidebar/stores/SidebarStore';

jest.mock('../registries/ViewRegistry', () => ({
    get: jest.fn(),
}));

jest.mock('../../Sidebar/stores/SidebarStore', () => ({
    clearConfig: jest.fn(),
}));

test('Render view returned from ViewRegistry', () => {
    viewRegistry.get.mockReturnValue(() => (<h1>Test</h1>));
    const view = render(<ViewRenderer router={{route: {view: 'test'}}} />);
    expect(view).toMatchSnapshot();
    expect(viewRegistry.get).toBeCalledWith('test');
    expect(sidebarStore.clearConfig).toBeCalled();
});

test('Render view returned from ViewRegistry with passed router', () => {
    const router = {
        route: {
            view: 'test',
        },
        attributes: {
            value: 'Test attribute',
        },
    };

    viewRegistry.get.mockReturnValue((props) => (<h1>{props.router.attributes.value}</h1>));
    const view = render(<ViewRenderer router={router} />);
    expect(view).toMatchSnapshot();
    expect(viewRegistry.get).toBeCalledWith('test');
});

test('Render view should throw if view does not exist', () => {
    viewRegistry.get.mockReturnValue(undefined);
    expect(() => render(<ViewRenderer router={{route: {view: 'not_existing'}}} />)).toThrow(/not_existing/);
});

test('Render view with parents should nest rendered views', () => {
    const router = {
        route: {
            name: 'sulu_admin.form_tab',
            view: 'form_tab',
            parent: {
                name: 'sulu_admin.form',
                view: 'form',
                parent: {
                    name: 'sulu_admin.app',
                    view: 'app',
                },
            },
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'form_tab':
                return function FormTab(props) {
                    return (
                        <div>
                            <h3>Form Tab</h3>
                            {props.route.name}
                        </div>
                    );
                };
            case 'form':
                return function Form(props) {
                    return (
                        <div>
                            <h2>Form</h2>
                            {props.route.name}
                            {props.children()}
                        </div>
                    );
                };
            case 'app':
                return function App(props) {
                    return (
                        <div>
                            <h1>App</h1>
                            {props.route.name}
                            {props.children()}
                        </div>
                    );
                };
        }
    });

    expect(render(<ViewRenderer router={router} />)).toMatchSnapshot();
});

test('Render view with parents should nest rendered views and correctly pass children arguments', () => {
    const router = {
        route: {
            name: 'sulu_admin.form_tab',
            view: 'form_tab',
            parent: {
                name: 'sulu_admin.form',
                view: 'form',
                parent: {
                    name: 'sulu_admin.app',
                    view: 'app',
                },
            },
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'form_tab':
                return function FormTab(props) {
                    return (
                        <div>
                            <h3>Form Tab</h3>
                            <p>{props.route.name}</p>
                            <p>{props.form}</p>
                        </div>
                    );
                };
            case 'form':
                return function Form(props) {
                    return (
                        <div>
                            <h2>Form</h2>
                            <p>{props.route.name}</p>
                            <p>{props.app}</p>
                            {props.children({form: 'Form'})}
                        </div>
                    );
                };
            case 'app':
                return function App(props) {
                    return (
                        <div>
                            <h1>App</h1>
                            <p>{props.route.name}</p>
                            {props.children({app: 'App'})}
                        </div>
                    );
                };
        }
    });

    expect(render(<ViewRenderer router={router} />)).toMatchSnapshot();
});

test('Render view with not existing parent should throw', () => {
    const router = {
        route: {
            view: 'form_tab',
            parent: {
                view: 'form',
                parent: {
                    view: 'app',
                },
            },
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'form_tab':
                return function FormTab() {
                    return (
                        <div>
                            <h3>Form Tab</h3>
                        </div>
                    );
                };
            case 'form':
                return function Form(props) {
                    return (
                        <div>
                            <h2>Form</h2>
                            {props.children}
                        </div>
                    );
                };
        }
    });

    expect(() => render(<ViewRenderer router={router} />)).toThrow(/app/);
});

test('Render view with route that has no rerenderAttributes', () => {
    const router = {
        route: {
            view: 'webspaceOverview',
        },
        attributes: {
            webspace: 'test',
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'webspaceOverview':
                return function WebspaceOverview() {
                    return (
                        <div>
                            <h3>Webspace</h3>
                        </div>
                    );
                };
        }
    });

    const viewRenderer = shallow(<ViewRenderer router={router} />);
    expect(viewRenderer.key()).toBe(undefined);
});

test('Render view with route that has rerenderAttributes', () => {
    const router = {
        route: {
            view: 'webspaceOverview',
            rerenderAttributes: [
                'webspace',
            ],
        },
        attributes: {
            webspace: 'test',
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'webspaceOverview':
                return function WebspaceOverview() {
                    return (
                        <div>
                            <h3>Webspace</h3>
                        </div>
                    );
                };
        }
    });

    const viewRenderer = shallow(<ViewRenderer router={router} />);
    expect(viewRenderer.key()).toBe('test');
});

test('Render view with route that has more than one rerenderAttributes', () => {
    const router = {
        route: {
            view: 'webspaceOverview',
            rerenderAttributes: [
                'webspace',
                'locale',
            ],
        },
        attributes: {
            webspace: 'test',
            locale: 'de',
        },
    };

    viewRegistry.get.mockImplementation((view) => {
        switch (view) {
            case 'webspaceOverview':
                return function WebspaceOverview() {
                    return (
                        <div>
                            <h3>Webspace</h3>
                        </div>
                    );
                };
        }
    });

    const viewRenderer = shallow(<ViewRenderer router={router} />);
    expect(viewRenderer.key()).toBe('test__de');
});

test('Render view and not clear the sidebarstore when component has sidebar', () => {
    const Component = class Component extends React.Component {
        static hasSidebar = true;

        render() {
            return <h1>{this.props.title}</h1>;
        }
    };

    viewRegistry.get.mockReturnValue(Component);
    const view = render(<ViewRenderer router={{route: {view: 'test'}}} />);
    expect(view).toMatchSnapshot();
    expect(viewRegistry.get).toBeCalledWith('test');
    expect(sidebarStore.clearConfig).not.toBeCalled();
});
