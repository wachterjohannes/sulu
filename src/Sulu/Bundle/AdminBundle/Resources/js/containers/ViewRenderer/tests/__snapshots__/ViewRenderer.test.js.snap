// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render view and not clear the sidebarstore when component has sidebar 1`] = `<h1 />`;

exports[`Render view returned from ViewRegistry 1`] = `
<h1>
  Test
</h1>
`;

exports[`Render view returned from ViewRegistry with passed router 1`] = `
<h1>
  Test attribute
</h1>
`;

exports[`Render view with parents should nest rendered views 1`] = `
<div>
  <h1>
    App
  </h1>
  sulu_admin.app
  <div>
    <h2>
      Form
    </h2>
    sulu_admin.form
    <div>
      <h3>
        Form Tab
      </h3>
      sulu_admin.form_tab
    </div>
  </div>
</div>
`;

exports[`Render view with parents should nest rendered views and correctly pass children arguments 1`] = `
<div>
  <h1>
    App
  </h1>
  <p>
    sulu_admin.app
  </p>
  <div>
    <h2>
      Form
    </h2>
    <p>
      sulu_admin.form
    </p>
    <p>
      App
    </p>
    <div>
      <h3>
        Form Tab
      </h3>
      <p>
        sulu_admin.form_tab
      </p>
      <p>
        Form
      </p>
    </div>
  </div>
</div>
`;
